dokka-subprojects/analysis-kotlin-descriptors-compiler/src/main/kotlin/org/jetbrains/dokka/analysis/kotlin/descriptors/compiler/impl/DescriptorSampleAnalysisEnvironment.kt

```diff
@@ -0,0 +1,195 @@
+
+    override fun <T> use(block: SampleAnalysisEnvironment.() -> T): T {
+        // Run from the thread of Dispatchers.Default as it can help
+        // avoid memory leaks through the compiler's ThreadLocals.
+        // Might not be relevant if the project stops using coroutines.
+        return runBlocking(Dispatchers.Default) {
+            @OptIn(DokkaPluginApiPreview::class)
+            SamplesKotlinAnalysis(
+                sourceSets = context.configuration.sourceSets,
+                context = context,
+                projectKotlinAnalysis = descriptorAnalysisPlugin.querySingle { kotlinAnalysis }
+            ).use { kotlinAnalysis ->
+                val sampleAnalysis = DescriptorSampleAnalysisEnvironment(
+                    kdocFinder = descriptorAnalysisPlugin.querySingle { kdocFinder },
+                    kotlinAnalysis = kotlinAnalysis,
+                    dokkaLogger = context.logger
+                )
+                block(sampleAnalysis)
+            }
+        }
+    }
+}
+
+internal class DescriptorSampleAnalysisEnvironment(
+    private val kdocFinder: KDocFinder,
+    private val kotlinAnalysis: KotlinAnalysis,
+    private val dokkaLogger: DokkaLogger,
+) : SampleAnalysisEnvironment {
+
+    override fun resolveSample(
+        sourceSet: DokkaConfiguration.DokkaSourceSet,
+        fullyQualifiedLink: String,
+    ): SampleSnippet? {
+        val resolveSession = kotlinAnalysis[sourceSet].resolveSession
+
+        val samplePsiElement = resolveSession.resolveSamplePsiElement(sourceSet, fullyQualifiedLink)
+        if (samplePsiElement == null) {
+            dokkaLogger.debug("Cannot resolve sample element for: \"$fullyQualifiedLink\"")
+            return null
+        } else if (samplePsiElement.containingFile !is KtFile) {
+            dokkaLogger.warn("Unable to resolve non-Kotlin @sample links: \"$fullyQualifiedLink\"")
+            return null
+        }
+
+        return SampleSnippet(
+            imports = processImports(samplePsiElement),
+            body = processBody(samplePsiElement)
+        )
+    }
+
+    private fun ResolveSession.resolveSamplePsiElement(
+        dokkaSourceSet: DokkaConfiguration.DokkaSourceSet,
+        fqLink: String,
+    ): PsiElement? {
+        val packageDescriptor = resolveNearestPackageDescriptor(fqLink)
+        if (packageDescriptor == null) {
+            dokkaLogger.debug(
+                "Unable to resolve package descriptor for @sample: \"$fqLink\";"
+            )
+            return null
+        }
+
+        val kdocLink = kdocFinder.resolveKDocLink(
+            fromDescriptor = packageDescriptor,
+            qualifiedName = fqLink,
+            sourceSet = dokkaSourceSet,
+            emptyBindingContext = true
+        ).firstOrNull()
+
+        if (kdocLink == null) {
+            dokkaLogger.warn(
+                "Unable to resolve a @sample link: \"$fqLink\". Is it used correctly? " +
```

Comment:
I missed in the first review the message does not have the location of sample link. It can be useful to find the link in the input code.